python/dgl/sampling/pinsage.py

```diff
@@ -0,0 +1,214 @@
+"""PinSAGE sampler & related functions and classes"""
+
+import numpy as np
+
+from .. import backend as F
+from .. import convert
+from .. import transform
+from .. import function as fn
+from .randomwalks import random_walk
+from .neighbor import select_topk
+from ..base import EID
+
+
+class GenericPinSAGESampler(object):
+    """PinSAGE-like sampler extended to any heterographs, given a metapath.
+
+    Given a heterogeneous graph, PinSAGE neighbor sampler would generate a homogeneous
+    graph where the neighbors of each node are the most commonly visited nodes of the
+    same type by random walk with restarts.  The random walk with restarts are based
+    on a given metapath, which should have the same beginning and ending node type.
+
+    Parameters
+    ----------
+    G : DGLHeteroGraph
+        The bidirectional bipartite graph.
+
+        The graph should only have two node types: ``ntype`` and ``other_type``.
+        The graph should only have two edge types, one connecting from ``ntype`` to
+        ``other_type``, and another connecting from ``other_type`` to ``ntype``.
+
+        PinSAGE works on a bidirectional bipartite graph where for each edge
+        going from node u to node v, there exists an edge going from node v to node u.
+    ntype : str
+        The node type for which the graph would be constructed on.
```

Comment:
Is this the same as the starting and the ending types of the given metapath?

Reply:
You're right.

Reply:
Then I guess this could be inferred from `metapath`?

Reply:
Removed.